File: apps/worker/app/crypto.py
from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def decrypt_api_key(encrypted: str, hex_key: str) -> str:
    """Decrypt AES-256-GCM encrypted API key from NestJS EncryptionService.

    Format: "{iv_hex}:{auth_tag_hex}:{ciphertext_hex}"
    IV = 12 bytes, auth tag = 16 bytes, all hex-encoded, colon-separated.
    """
    parts = encrypted.split(":")
    if len(parts) != 3:
        raise ValueError(
            f"Invalid encrypted key format: expected 3 colon-separated parts, got {len(parts)}"
        )

    iv_hex, auth_tag_hex, ciphertext_hex = parts
    iv = bytes.fromhex(iv_hex)
    auth_tag = bytes.fromhex(auth_tag_hex)
    ciphertext = bytes.fromhex(ciphertext_hex)
    key = bytes.fromhex(hex_key)

    # AESGCM expects ciphertext + auth_tag concatenated
    aesgcm = AESGCM(key)
    plaintext = aesgcm.decrypt(iv, ciphertext + auth_tag, None)
    return plaintext.decode("utf-8")

File: apps/gateway/src/crypto.ts
import { createDecipheriv } from 'node:crypto';

const ALGORITHM = 'aes-256-gcm';

/**
 * Decrypt an AES-256-GCM ciphertext stored by the NestJS EncryptionService.
 * Format: `{ivHex}:{tagHex}:{ciphertextHex}`
 */
export function decryptApiKey(encrypted: string, hexKey: string): string {
  const parts = encrypted.split(':');
  if (parts.length !== 3) throw new Error('Invalid encrypted key format');
  const [ivHex, tagHex, dataHex] = parts as [string, string, string];

  const key = Buffer.from(hexKey, 'hex');
  const iv = Buffer.from(ivHex, 'hex');
  const tag = Buffer.from(tagHex, 'hex');
  const data = Buffer.from(dataHex, 'hex');

  const decipher = createDecipheriv(ALGORITHM, key, iv);
  decipher.setAuthTag(tag);
  return decipher.update(data).toString('utf8') + decipher.final('utf8');
}
